Improvements to Gradient2D Editor

MewPurPur opened the pull request:
A few small tweaks to the GradientTexture2D editor.

- Modifies snap_size into snap_count, which everyone I asked seemed to prefer for Gradient in #70760. I think it does make more sense here too, as we're representing fractions and not pixels. But feel free to discuss
![image](https://user-images.githubusercontent.com/85438892/210734238-53f64786-e961-4e4e-a97d-862378c5c37b.png)
- If two handles are overlapping and the user clicks the area where they overlap, picks the handle that's closer, instead of always the FROM handle.
- Remembers snapping options between sessions on a per-GradientTexture2D basis.

**Improvements to Undo/Redo**
- Only leaves a message in the output when you release a handle to prevent spam
- Fixes occasions in which it used to break
- Doesn't leave a message if you leave the handle where you grabbed it

All of these Undo/Redo behaviors match how it is done in the CanvasItem editor.

MewPurPur committed:
Improvements to Gradient2D Editor

MewPurPur commented:
This hasn't been receiving much attention, so might as well finish the job of polishing it.

- Added hover indicators for handles
- Added axis snapping with Shift
- Added temporary grid snapping with Ctrl

These are pretty expected functionalities from a 2D editor.

KoBeWi commented:
> Added temporary grid snapping with Ctrl

I think the editor should update as soon as you press Ctrl. Right now it has no effect until you drag.

KoBeWi commented on editor/plugins/gradient_texture_2d_editor_plugin.cpp:
```diff
@@ -33,79 +33,136 @@
-Point2 GradientTexture2DEditorRect::_get_handle_position(const Handle p_handle) {
+void GradientTexture2DEdit::_on_mouse_exited() {
+	hovered = HANDLE_NONE;
+	queue_redraw();
```
This redraw isn't always needed.
(not a big problem tho)

KoBeWi commented on editor/plugins/gradient_texture_2d_editor_plugin.cpp:
```diff
@@ -33,79 +33,136 @@
 
-	String property_name = handle == HANDLE_FILL_FROM ? "fill_from" : "fill_to";
-
+	String property_name = (grabbed == HANDLE_FROM) ? "fill_from" : "fill_to";
```
```suggestion
	const StringName property_name = (grabbed == HANDLE_FROM) ? "fill_from" : "fill_to";
```

KoBeWi commented on editor/plugins/gradient_texture_2d_editor_plugin.cpp:
```diff
@@ -33,79 +33,136 @@
+	snap_count = p_snap_count;
 	queue_redraw();
+	if (texture.is_valid()) {
+		if (snap_count != 10) {
```
`10` could be a constant (`static const int`), as it's used more than once.

KoBeWi commented on editor/plugins/gradient_texture_2d_editor_plugin.cpp:
```diff
@@ -134,48 +191,28 @@ void GradientTexture2DEditorRect::_notification(int p_what) {
 			// Draw handles.
-			draw_texture(fill_from_icon, (_get_handle_position(HANDLE_FILL_FROM) - handle_size / 2).round());
-			draw_texture(fill_to_icon, (_get_handle_position(HANDLE_FILL_TO) - handle_size / 2).round());
+			Color focus_modulate = Color(0.5, 1, 2);
```
```suggestion
			const Color focus_modulate = Color(0.5, 1, 2);
```

KoBeWi commented on editor/plugins/gradient_texture_2d_editor_plugin.cpp:
```diff
@@ -33,81 +33,140 @@
 #include "editor/editor_node.h"
 #include "editor/editor_scale.h"
 #include "editor/editor_undo_redo_manager.h"
-#include "editor/gui/editor_spin_slider.h"
```
Looks like you deleted it in a wrong file.

MewPurPur replied:
Ah, true! I was a bit confused what you meant and was waiting for it to compile... now I see XD

MewPurPur replied:
Though I don't see anything like this in the header, huh. I did do `class EditorSpinSlider` twice so I removed that tho.

MewPurPur commented:
A bit short on time, sorry, but I think I addressed most of the remarks. Just not the static variable one, I'm not familiar with this feature, sorry-

KoBeWi commented:
Example is here:
https://github.com/godotengine/godot/blob/14c582bca81046fdde35e16088ddfd5df0136d56/scene/resources/packed_scene.h#L224
It's `static const`, so that it's a constant that's allocated only once for the whole class (without `static` it's once per instance).
You need to initialize it in cpp file.

MewPurPur commented:
> I think the editor should update as soon as you press Ctrl. Right now it has no effect until you drag.

I removed the implementation, I actually found it very odd when the grid appeared when you do stuff like Ctrl+Z, so I think it is reasonable to only show when you are dragging.

Implement the static const thing too. Also moved drawing to a `_draw()` function.

KoBeWi commented:
Can't disable snapping.

https://user-images.githubusercontent.com/2223172/236692823-26c16421-f571-423a-b031-b4b9a437068f.mp4

MewPurPur commented:
Ah thanks. No idea what could possibly be causing this, but I'll try. The only thing relating to those functions I did was using the meta suggestions...

MewPurPur commented:
The GradientTexture2D now gets rebuilt every time I press the snap button. I'm 90% sure this is because the added editor metadata causes the property tree to be updated, rebuilding this editor plugin, which I shouldn't happen.

KoBeWi commented:
#76814

MewPurPur commented:
Ah that's true. Thanks for pointing it out. I just saw that the code for the "READY" notif was running and rebuilding the editor, so I thought this might be the problem.

KoBeWi commented:
> I removed the implementation, I actually found it very odd when the grid appeared when you do stuff like Ctrl+Z, so I think it is reasonable to only show when you are dragging.

But it could at least update when you press Ctrl while dragging (without any movement). That's what I meant.

Calinou commented:
> But it could at least update when you press Ctrl while dragging (without any movement). That's what I meant.

Note that this also needs to be done in the curve editor PR.

Calinou reviewed (approved):
Tested locally (rebased against latest `master` 06ccbfe8eec6e840db3df2360dc2804386f940b8), it works.

Code looks good to me at a glance.

akien-mga commented:
Thanks!
